sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala

```diff
@@ -594,23 +596,29 @@ object IntervalUtils {
       s"$fallbackNotice")
     var micros: Long = 0L
     var days: Int = 0
+    val sign = if (m.group("sign") != null && m.group("sign") == "-") -1 else 1
     unitsRange(to, from).foreach {
       case unit @ DAY =>
-        days = toLongWithRange(unit, m.group(unit.toString), 0, Int.MaxValue).toInt
+        days = sign * toLongWithRange(unit, m.group(unit.toString), 0, Int.MaxValue).toInt
       case unit @ HOUR =>
-        val parsed = toLongWithRange(unit, m.group(unit.toString), 0, 23)
-        micros = Math.addExact(micros, parsed * MICROS_PER_HOUR)
+        val parsed = toLongWithRange(unit, m.group(unit.toString), 0,
+          if (from == HOUR) 2562047788L else 23)
```

Comment:
Let's add constants for the magic number 2562047788L and for 153722867280L. It might be not clear what they mean for unfamiliar devs.